{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook shows how to prepare custom genome coordinate for versions/species that is not currently supported. Or to integrate outside database information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode = BedTool('/projects/ps-yeolab3/bay001/annotations/ChlSab2/Chlorocebus_sabaeus.ChlSab1.1.102.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# these functions may need to be modified depending on what file format your are inputting\n",
    "  \n",
    "import pickle\n",
    "\n",
    "def first_last_exon_cds(all_dict):\n",
    "    ''' find first last exon and cds'''\n",
    "    \n",
    "    for key in all_dict.keys():\n",
    "        features = all_dict[key]['features']\n",
    "        \n",
    "        #### exon\n",
    "        if 'exon' in features.keys():\n",
    "            min_start = min([e[0] for e in list(features['exon'])])\n",
    "            max_start = max([e[1] for e in list(features['exon'])])\n",
    "        \n",
    "            if all_dict[key]['strand'] == '+':\n",
    "                features['first_exon'] = set([e for e in list(features['exon']) if e[0] == min_start])\n",
    "                features['last_exon'] = set([e for e in list(features['exon']) if e[1] == max_start])\n",
    "            else:\n",
    "                features['last_exon'] = set([e for e in list(features['exon']) if e[0] == min_start])\n",
    "                features['first_exon'] = set([e for e in list(features['exon']) if e[1] == max_start])\n",
    "        \n",
    "            features['exon'] = features['exon'] - features['first_exon'] - features['last_exon']\n",
    "        else:\n",
    "            #del all_dict[key] # no exon, no need to keep\n",
    "            pass\n",
    "        ## CDS\n",
    "        if 'CDS' in features.keys():\n",
    "            min_start = min([e[0] for e in list(features['CDS'])])\n",
    "            max_start = max([e[1] for e in list(features['CDS'])])\n",
    "        \n",
    "            if all_dict[key]['strand'] == '+':\n",
    "                features['first_CDS'] = set([e for e in list(features['CDS']) if e[0] == min_start])\n",
    "                features['last_CDS'] = set([e for e in list(features['CDS']) if e[1] == max_start])\n",
    "            else:\n",
    "                features['last_CDS'] = set([e for e in list(features['CDS']) if e[0] == min_start])\n",
    "                features['first_CDS'] = set([e for e in list(features['CDS']) if e[1] == max_start])\n",
    "        \n",
    "            features['CDS'] = features['CDS'] - features['first_CDS'] - features['last_CDS']\n",
    "        \n",
    "def five_three_utr(all_dict):\n",
    "    ''' hg19 annotations don't have UTR'''\n",
    "    for key in all_dict.keys():\n",
    "        features = all_dict[key]['features']\n",
    "        \n",
    "        #### exon\n",
    "        if 'UTR' in features.keys():\n",
    "            min_start = min([e[0] for e in list(features['UTR'])])\n",
    "            max_start = max([e[1] for e in list(features['UTR'])])\n",
    "        \n",
    "            if all_dict[key]['strand'] == '+':\n",
    "                features['five_prime_UTR'] = set([e for e in list(features['UTR']) if e[0] == min_start])\n",
    "                features['three_prime_UTR'] = set([e for e in list(features['UTR']) if e[1] == max_start])\n",
    "            else:\n",
    "                features['five_prime_UTR'] = set([e for e in list(features['UTR']) if e[1] == max_start])\n",
    "                features['three_prime_UTR'] = set([e for e in list(features['UTR']) if e[0] == min_start])\n",
    "        \n",
    "            del features['UTR']\n",
    "\n",
    "\n",
    "        \n",
    "def build_transcript_dict(gencode = gencode, outdir = '/home/hsher/Metadensity/metadensity/data/ChlSab'):\n",
    "    ''' extract gencode coordinate and save in data'''\n",
    "    \n",
    "    # make a directory for every coordinate\n",
    "    annotation_path = outdir\n",
    "    if not os.path.exists(annotation_path):\n",
    "        os.mkdir(annotation_path)\n",
    "    \n",
    "    all_dict = {}\n",
    "    \n",
    "    gene = gencode.filter(lambda x: x[2] == 'gene')\n",
    "    transcript = gencode.filter(lambda x: x[2] == 'transcript')\n",
    "    exons = gencode.filter(lambda x: x[2] == 'exon')\n",
    "    intron = transcript.subtract(exons, s = True).saveas()\n",
    "    print('extracting from gencode annotations')\n",
    "    for g in gencode:\n",
    "        feature_type = g[2]\n",
    "        \n",
    "        if feature_type == 'gene':\n",
    "            # start a new dict\n",
    "            all_dict[g.attrs['gene_id']] = {} # ENST or ENSG\n",
    "            all_dict[g.attrs['gene_id']]['chrom'] = 'chr'+g.chrom # NEED TO DO THE SAME FOR GENES/TRANSCRIPTS TOO; NEED TO CONSISTENT WITH CLIP PIPELINE\n",
    "            all_dict[g.attrs['gene_id']]['start'] = g.start\n",
    "            all_dict[g.attrs['gene_id']]['end'] = g.end\n",
    "            all_dict[g.attrs['gene_id']]['strand'] = g.strand\n",
    "            all_dict[g.attrs['gene_id']]['id'] = g.attrs['gene_id']\n",
    "            all_dict[g.attrs['gene_id']]['type'] = g.attrs['gene_biotype']\n",
    "            try:\n",
    "                all_dict[g.attrs['gene_id']]['name'] = g.attrs['gene_name']\n",
    "            except:\n",
    "                all_dict[g.attrs['gene_id']]['name'] = ''\n",
    "            all_dict[g.attrs['gene_id']]['features'] = {} # start to contain stuffs\n",
    "        else:\n",
    "             # doesn't always equal to ID, for X,Y chromosome genes\n",
    "            gene_id = g.attrs['gene_id']\n",
    "            \n",
    "            for ids in [gene_id]:\n",
    "                target_dict = all_dict[ids] ############## YOU NEED TO MAKE SURE THE FEATURE NAMES ARE IN METAGENE ###########\n",
    "                if feature_type == 'five_prime_utr':\n",
    "                    feature_type = 'five_prime_UTR'\n",
    "                if feature_type == 'three_prime_utr':\n",
    "                    feature_type = 'three_prime_UTR'\n",
    "                if feature_type not in target_dict['features'].keys():\n",
    "                    target_dict['features'][feature_type] = set()\n",
    "                target_dict['features'][feature_type].update([(g.start, g.end)])\n",
    "    print('building intron')\n",
    "    for i in intron:\n",
    "        feature_type = 'intron'\n",
    "        \n",
    "        gene_id = i.attrs['gene_id']\n",
    "            \n",
    "        for ids in [gene_id]:\n",
    "            target_dict = all_dict[ids]\n",
    "            if feature_type not in target_dict['features'].keys():\n",
    "                target_dict['features'][feature_type] = set()\n",
    "            target_dict['features'][feature_type].update([(i.start, i.end)])\n",
    "    \n",
    "    print('building first last exon/cds')\n",
    "    first_last_exon_cds(all_dict)\n",
    "    print('building five/three prime UTR')\n",
    "    five_three_utr(all_dict)\n",
    "    print('writing to directory')\n",
    "    with open(os.path.join(annotation_path, 'gencode'), 'wb') as f:\n",
    "        pickle.dump(all_dict, f)\n",
    "    return all_dict\n",
    "            \n",
    "            \n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "extracting from gencode annotations\n",
      "building intron\n",
      "building first last exon/cds\n",
      "building five/three prime UTR\n",
      "writing to directory\n"
     ]
    }
   ],
   "source": [
    "d = build_transcript_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chrom': 'chr8',\n",
       " 'start': 183179,\n",
       " 'end': 246703,\n",
       " 'strand': '+',\n",
       " 'id': 'ENSCSAG00000017073',\n",
       " 'type': 'protein_coding',\n",
       " 'name': 'FBXO25',\n",
       " 'features': {'transcript': {(183179, 246703)},\n",
       "  'exon': {(184546, 184671),\n",
       "   (189275, 189416),\n",
       "   (209060, 209164),\n",
       "   (210598, 210648),\n",
       "   (213557, 213650),\n",
       "   (228044, 228138),\n",
       "   (229299, 229484),\n",
       "   (236307, 236490),\n",
       "   (240928, 241072),\n",
       "   (245629, 245656)},\n",
       "  'CDS': {(209060, 209164),\n",
       "   (210598, 210648),\n",
       "   (213557, 213650),\n",
       "   (228044, 228138),\n",
       "   (229299, 229484),\n",
       "   (236307, 236490),\n",
       "   (240928, 241072),\n",
       "   (245629, 245656)},\n",
       "  'start_codon': {(189282, 189285)},\n",
       "  'five_prime_UTR': {(183179, 183285), (184546, 184671), (189275, 189282)},\n",
       "  'intron': {(183285, 184546),\n",
       "   (184671, 189275),\n",
       "   (189416, 209060),\n",
       "   (209164, 210598),\n",
       "   (210648, 213557),\n",
       "   (213650, 228044),\n",
       "   (228138, 229299),\n",
       "   (229484, 236307),\n",
       "   (236490, 240928),\n",
       "   (241072, 245629),\n",
       "   (245656, 246616)},\n",
       "  'first_exon': {(183179, 183285)},\n",
       "  'last_exon': {(246616, 246703)},\n",
       "  'first_CDS': {(189282, 189416)},\n",
       "  'last_CDS': {(246616, 246703)}}}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d['ENSCSAG00000017073']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# And you also need to edit the `metadensity.config file`.\n",
    "\n",
    "adding this:\n",
    "\n",
    "```'ChlSab':{\n",
    "                'gencode': 'Chlor.gtf',\n",
    "                'transcript': 'Chlor.genes.gtf',\n",
    "                'gencode_feature': 'Chlor.combined.sorted.gtf',\n",
    "                ```\n",
    "                \n",
    "to self.file_dict\n",
    "\n",
    "\n",
    "these files are processed by `/home/hsher/project/covid-clip/notebook/parse_coord.sh`\n",
    "just to seperate the genes, and add chr to the front.\n",
    "\n",
    "code:\n",
    "```\n",
    "awk '{ if ($3 == \"gene\"){ print } }' /projects/ps-yeolab3/bay001/annotations/ChlSab2/Chlorocebus_sabaeus_Sars_cov_2.gtf > ~/gencode_coords/Chlor.genes.gtf\n",
    "\n",
    "sed -i -e 's/^/chr/' ~/gencode_coords/Chlor.genes.gtf\n",
    "\n",
    "awk '{ if ($3 != \"gene\" && $3 != \"transcript\"){ print } }' /projects/ps-yeolab3/bay001/annotations/ChlSab2/Chlorocebus_sabaeus_Sars_cov_2.gtf > ~/gencode_coords/Chlor.combined.sorted.gtf\n",
    "\n",
    "sed -i -e 's/^/chr/' ~/gencode_coords/Chlor.combined.sorted.gtf\n",
    "```\n",
    "\n",
    "Then reinstall!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
